internal/state/graph/gateway.go

```diff
@@ -134,278 +135,304 @@ func buildListeners(
+					)
+					return []conditions.Condition{conditions.NewListenerUnsupportedProtocol(valErr.Error())}
+				},
+			},
+		},
+		http: &listenerConfigurator{
+			validators: []listenerValidator{
+				validateListenerHostname,
+				createAddressesValidator(gw),
+				createNoValidGatewayClassValidator(gc),
+				validateHTTPListener,
+			},
+			conflictResolvers: []listenerConflictResolver{
+				createHostnameConflictResolver(),
+			},
+		},
+		https: &listenerConfigurator{
+			validators: []listenerValidator{
+				validateListenerHostname,
+				createAddressesValidator(gw),
+				createNoValidGatewayClassValidator(gc),
+				createHTTPSListenerValidator(gw.Namespace),
+			},
+			conflictResolvers: []listenerConflictResolver{
+				createHostnameConflictResolver(),
+			},
+			externalReferenceResolvers: []listenerExternalReferenceResolver{
+				createExternalReferencesForTLSSecretsResolver(gw.Namespace, secretMemoryMgr),
+			},
+		},
 	}
 }
 
-type httpListenerConfigurator struct {
-	gateway         *v1beta1.Gateway
-	secretMemoryMgr secrets.SecretDiskMemoryManager
-	usedHostnames   map[string]*Listener
-	validate        func(gl v1beta1.Listener) []conditions.Condition
+// listenerValidator validates a listener. If the listener is invalid, the validator will return appropriate conditions.
+type listenerValidator func(v1beta1.Listener) []conditions.Condition
+
+// listenerConflictResolver resolves conflicts between listeners. In case of a conflict, the resolver will make
+// the conflicting listeners invalid - i.e. it will modify the passed listener and the previously processed conflicting
+// listener. It will also add appropriate conditions to the listeners.
+type listenerConflictResolver func(listener *Listener)
+
+// listenerExternalReferenceResolver resolves external references for a listener. If the reference is not resolvable,
+// the resolver will make the listener invalid and add appropriate conditions.
+type listenerExternalReferenceResolver func(listener *Listener)
+
+// listenerConfigurator is responsible for configuring a listener.
+// validators, conflictResolvers, externalReferenceResolvers generate conditions for invalid fields of the listener.
+// Because the Gateway status includes a status field for each listener, the messages in those conditions
+// don't need to include the full path to the field (e.g. "spec.listeners[0].hostname"). They will include
+// a path starting from the field of a listener (e.g. "hostname", "tls.options").
+type listenerConfigurator struct {
+	// validators must not depend on the order of execution.
+	validators []listenerValidator
+
+	// conflictResolvers can depend on validators - they will only be executed if all validators pass.
+	conflictResolvers []listenerConflictResolver
+	// externalReferenceResolvers can depend on validators - they will only be executed if all validators pass.
+	externalReferenceResolvers []listenerExternalReferenceResolver
```

Comment:
why slice?

implementing support for https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.AllowedRoutes can go here. For example, what if the namespace selector resolves into 0 namespaces - this way we will be able to catch it and handle it here.